### PR TITLE
add sorting by position in a blob file

### PR DESCRIPTION
**Pull request by s-mx**
before we were splitting chunks by index order but it is inefficient
because if random reads. now we read whole index, sort it by position
and split chunk in that order.

### CODE REVIEW

File: tests/unit/defrag.cpp [hunk 1]
```diff
@@ -576,3 +576,29 @@ BOOST_AUTO_TEST_CASE(test_defrag_in_dir) {
 	// check that defrag succeeded
 	BOOST_REQUIRE_EQUAL(eblob_stat_get(backend->stat, EBLOB_GST_DATASORT_COMPLETION_STATUS), 0);
 }
+
+
+BOOST_AUTO_TEST_CASE(test_defrag_by_position) {
+	eblob_config_test_wrapper config_wrapper(true, EBLOB_LOG_INFO);
+	config_wrapper.config.records_in_blob = 1000000; // 10^5
+	config_wrapper.config.blob_size = 10000 * 10000; // 10^8
+
```
Comment (s-mx) [inline tests/unit/defrag.cpp:628]: add EBLOB_SORT_BY_POS_FLAG

File: tests/unit/defrag.cpp [hunk 2]
```diff
@@ -576,3 +576,29 @@ BOOST_AUTO_TEST_CASE(test_defrag_in_dir) {
 	// check that defrag succeeded
 	BOOST_REQUIRE_EQUAL(eblob_stat_get(backend->stat, EBLOB_GST_DATASORT_COMPLETION_STATUS), 0);
 }
+
+
+BOOST_AUTO_TEST_CASE(test_defrag_by_position) {
+	eblob_config_test_wrapper config_wrapper(true, EBLOB_LOG_INFO);
+	config_wrapper.config.records_in_blob = 1000000; // 10^5
```
Comment (s-mx) [inline tests/unit/defrag.cpp]: 10^6

File: library/blob.c [hunk 3]
```diff
@@ -1057,6 +1057,84 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
...
+static int eblob_blob_iterate_by_pos(struct eblob_iterate_priv *iter_priv)
+{
+	struct eblob_iterate_control *ctl = iter_priv->ctl;
+	struct eblob_base_ctl *bctl = ctl->base;
+
+	struct eblob_disk_control *dc = NULL;
+	struct eblob_iterate_local loc;
+	int err = 0;
+
+	static const int hdr = sizeof(struct eblob_disk_control);
```
Comment (sabramkin) [inline library/blob.c]: `size_t hdr_size`?

File: library/blob.c [hunk 4]
```diff
@@ -1057,6 +1057,84 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
 	return err;
 }
 
+
+/**
+ * eblob_disk_control_cmp_by_position() - compares dc's by position in a blob file
+ */
+static int eblob_disk_control_cmp_by_position(const void *lhs, const void *rhs) {
+	const struct eblob_disk_control *left = (const struct eblob_disk_control *)lhs;
```
Comment (sabramkin) [inline library/blob.c:1059]: Not sure, in C we can do implicit upcast from `void *` (`const struct eblob_disk_control *left = lhs;`)

File: tests/unit/defrag.cpp (hunk 2 above)
Comment (sabramkin) [reply tests/unit/defrag.cpp]: `int(1E6)` an no comment required =)

File: library/blob.c [hunk 5]
```diff
@@ -1057,6 +1057,84 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
...
+	dc = malloc(bctl->index_ctl.size);
+	if (dc == NULL) {
+		err = -ENOMEM;
+		goto err_out_check;
+	}
+
+	pthread_mutex_lock(&bctl->lock);
+	err = __eblob_read_ll(bctl->index_ctl.fd, dc, bctl->index_ctl.size, 0);
+	if (err) {
+		pthread_mutex_unlock(&bctl->lock);
```
Comment (sabramkin) [inline library/blob.c]: You unlock independently on `err`, so could do it once, before if.

File: tests/unit/defrag.cpp (hunk 2 above)
Comment (s-mx) [reply tests/unit/defrag.cpp]: Arguable.

File: library/blob.c (hunk 3 above)
Comment (s-mx) [reply library/blob.c]: inconsistent with other places. I would name it header_size in all places.

File: library/blob.c (hunk 5 above)
Comment (s-mx) [reply library/blob.c]: reasonable

Comment (s-mx) [reply library/blob.c]: done

File: library/blob.c (hunk 4 above)
Comment (s-mx) [reply library/blob.c:1059]: we can

File: tests/unit/eblob_wrapper.h [hunk 6]
```diff
@@ -168,6 +168,7 @@ make_default_item_generator(eblob_wrapper &wrapper, uint64_t seed = DEFAULT_RAND
 	return item_generator<std::piecewise_constant_distribution<double>>(wrapper, dist, seed);
 }
 
+<<<<<<< HEAD
```
Comment (s-mx) [inline tests/unit/eblob_wrapper.h]: bad

Comment (s-mx) [reply tests/unit/eblob_wrapper.h]: fixed

File: library/blob.c [hunk 7]
```diff
@@ -1099,7 +1172,17 @@ int eblob_blob_iterate(struct eblob_iterate_control *ctl)
 		}
 	}
 
-	err = eblob_blob_iterator(&iter_priv);
+	if (!(ctl->flags & EBLOB_ITERATE_FLAGS_BY_POSITION))
+		err = eblob_blob_iterator(&iter_priv);
+
+	if (ctl->flags & EBLOB_ITERATE_FLAGS_BY_POSITION) {
```
Comment (antmat) [inline library/blob.c]: maybe just `else` because the condition is the opposite one?

File: library/blob.c [hunk 8]
```diff
@@ -1057,6 +1053,83 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
...
 }
 
+
+/**
+ * eblob_disk_control_cmp_by_position() - compares dc's by position in a blob file
+ */
+static int eblob_disk_control_cmp_by_position(const void *lhs, const void *rhs) {
+	const struct eblob_disk_control *left = (const struct eblob_disk_control *)lhs;
+	const struct eblob_disk_control *right = (const struct eblob_disk_control *)rhs;
+	if (left->position == right->position)
```
Comment (antmat) [inline library/blob.c]: This function will be called many times, so I would start with more common cases, e.g.
```
if (left->position > right->position) {
    return 1;
} else if (left->position < right->position) {
    return -1;
} else {
    return 0;
}
```

File: library/blob.c [hunk 9]
```diff
@@ -1057,6 +1053,83 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
...
+	struct eblob_iterate_control *ctl = iter_priv->ctl;
+	struct eblob_base_ctl *bctl = ctl->base;
+
+	struct eblob_disk_control *dc = NULL;
+	struct eblob_iterate_local loc;
+	int err = 0;
+
+	static const int hdr = sizeof(struct eblob_disk_control);
+
+	if (iter_priv == NULL) {
```
Comment (antmat) [inline library/blob.c:1083]: herre you have check after usage on line 1074

File: library/blob.c (hunk 3 above)
Comment (antmat) [reply library/blob.c]: hdr_size is also used in blob.c:874, so I'd use more understandable name.

File: library/blob.c [hunk 10]
```diff
@@ -1057,6 +1053,83 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
...
+
+/**
+ * eblob_blob_iterate_by_pos() - read whole index, sort it by position and run iteration over it.
+ */
+static int eblob_blob_iterate_by_pos(struct eblob_iterate_priv *iter_priv)
+{
+	struct eblob_iterate_control *ctl = iter_priv->ctl;
+	struct eblob_base_ctl *bctl = ctl->base;
+
+	struct eblob_disk_control *dc = NULL;
```
Comment (antmat) [inline library/blob.c]: wecould use loc->dc instead of separate var, but it's up to you.

File: library/blob.c [hunk 11]
```diff
@@ -1057,6 +1053,83 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
...
+	err = eblob_check_disk(&loc);
+	eblob_bctl_release(bctl);
+
+err_out_check:
+	free(dc);
+	eblob_log(ctl->log, err < 0 ? EBLOB_LOG_ERROR : EBLOB_LOG_INFO, "blob-0.%d: iterated: data_fd: %d, index_fd: %d, "
+									"data_size: %llu, index_offset: %llu, err: %d\n",
+		  bctl->index, bctl->data_ctl.fd, bctl->index_ctl.fd, ctl->data_size, ctl->index_offset, err);
+
+	/*
```
Comment (antmat) [inline library/blob.c]: As we don;t use threads anymore this comment looks outdated

File: tests/unit/defrag.cpp [hunk 12]
```diff
@@ -566,3 +581,23 @@ BOOST_AUTO_TEST_CASE(test_defrag_in_dir) {
...
+	config_wrapper.config.records_in_blob = TOTAL_RECORDS;
+	config_wrapper.config.blob_flags |= EBLOB_ITERATE_FLAGS_BY_POSITION;
+
+	eblob_wrapper wrapper(config_wrapper.config);
+	std::vector<item_t> shadow_elems;
+	auto generator = make_uniform_item_generator(wrapper, 10, 10);
+	fill_eblob(wrapper, shadow_elems, generator, TOTAL_RECORDS);
+
+	eblob_base_ctl *first_bctl = container_of(wrapper.get()->bases.next, eblob_base_ctl, base_entry);
+	BOOST_REQUIRE_EQUAL(eblob_generate_sorted_index(wrapper.get(), first_bctl), 0);
```
Comment (antmat) [inline tests/unit/defrag.cpp:635]: maybe we could use autosort flag here

**Review by antmat — CHANGES_REQUESTED** [review]

File: tests/unit/defrag.cpp (hunk 12 above)
Comment (s-mx) [reply tests/unit/defrag.cpp:635]: Autosort flags proceeds in defrag thread, but all threads are disabled in unit tests

File: library/blob.c (hunk 3 above)
Comment (s-mx) [reply library/blob.c]: renamed

File: library/blob.c (hunk 7 above)
Comment (s-mx) [reply library/blob.c]: done

File: library/blob.c (hunk 8 above)
Comment (s-mx) [reply library/blob.c]: done with slightly different order

File: library/blob.c (hunk 9 above)
Comment (s-mx) [reply library/blob.c:1083]: fixed

File: library/blob.c (hunk 10 above)
Comment (s-mx) [reply library/blob.c]: done

File: library/blob.c (hunk 11 above)
Comment (s-mx) [reply library/blob.c]: removed

File: include/eblob/blob.h [hunk 13]
```diff
@@ -315,6 +315,11 @@ static inline void eblob_convert_disk_control(struct eblob_disk_control *ctl)
  */
 #define EBLOB_USE_VIEWS			(1<<12)
 
+/*
+ * Enable chunk splitting by position in a blob while defrag
+ */
+#define EBLOB_SORT_BY_POS		(1<<13)
```
Comment (antmat) [inline include/eblob/blob.h]: please check alignment

File: include/eblob/blob.h [hunk 14]
```diff
@@ -494,6 +499,7 @@ struct eblob_iterate_callbacks {
 #define EBLOB_ITERATE_FLAGS_READONLY		(1<<1)	/* do not modify entries while iterating a blob */
 #define EBLOB_ITERATE_FLAGS_INITIAL_LOAD	(1<<2)	/* set on initial load */
 #define EBLOB_ITERATE_FLAGS_VERIFY_CHECKSUM	(1<<3)	/* verify checksum for entries while iterating a blob */
+#define EBLOB_ITERATE_FLAGS_BY_POSITION		(1<<4)	/* iterate over entries in ascending order by position */
```
Comment (antmat) [inline include/eblob/blob.h]: please check alignment here too

File: include/eblob/blob.h [hunk 15]
```diff
@@ -879,6 +885,7 @@ static inline const char *eblob_dump_blob_flags(unsigned int flags) {
 		{ EBLOB_DISABLE_THREADS,		"disabled_threads"},
 		{ EBLOB_AUTO_INDEXSORT,			"auto_indexsort"},
 		{ EBLOB_USE_VIEWS,			"use_views"},
+		{ EBLOB_SORT_BY_POS,			"sort_by_pos"},
```
Comment (antmat) [inline include/eblob/blob.h:894]: is buffer still big enough to hold all possible flags?

**Review by antmat — CHANGES_REQUESTED** [review]

File: include/eblob/blob.h (hunk 15 above)
Comment (s-mx) [reply include/eblob/blob.h:894]: the length of all flags is 178 but I don't know how much we need at eblob_dump_flags_raw.

File: include/eblob/blob.h (hunk 13 above)
Comment (s-mx) [reply include/eblob/blob.h]: done

File: include/eblob/blob.h (hunk 14 above)
Comment (s-mx) [reply include/eblob/blob.h]: done

File: include/eblob/blob.h (hunk 15 above)
Comment (antmat) [reply include/eblob/blob.h:894]: Let's increase it to 512 just in case

File: library/blob.c [hunk 16]
```diff
@@ -1057,6 +1053,82 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
...
+
+	memset(&loc, 0, sizeof(loc));
+
+	loc.iter_priv = iter_priv;
+	loc.index_offset = 0;
+	loc.pos = 0;
+	loc.num = bctl->index_ctl.size / hdr_size;
+
+	loc.dc = malloc(bctl->index_ctl.size);
+	if (loc.dc == NULL) {
```
Comment (antmat) [inline library/blob.c]: minor: maybe we could used same rules as we agreed on elliptics not to check for ENOMEM?

File: library/blob.c [hunk 17]
```diff
@@ -1057,6 +1053,82 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
...
+	bctl = ctl->base;
+
+	memset(&loc, 0, sizeof(loc));
+
+	loc.iter_priv = iter_priv;
+	loc.index_offset = 0;
+	loc.pos = 0;
+	loc.num = bctl->index_ctl.size / hdr_size;
+
+	loc.dc = malloc(bctl->index_ctl.size);
```
Comment (shaitan) [inline library/blob.c]:
```suggestion
	loc.dc = calloc(loc.num, hdr_size);
```

File: library/blob.c [hunk 18]
```diff
@@ -1057,6 +1053,82 @@ static int eblob_blob_iterator(struct eblob_iterate_priv *iter_priv)
...
+	if (loc.dc == NULL) {
+		err = -ENOMEM;
+		goto err_out_check;
+	}
+
+	pthread_mutex_lock(&bctl->lock);
+	err = __eblob_read_ll(bctl->index_ctl.fd, loc.dc, bctl->index_ctl.size, 0);
```
Comment (shaitan) [inline library/blob.c]:
```suggestion
	err = __eblob_read_ll(bctl->index_ctl.fd, loc.dc, loc.num * hdr_size, 0);
```

Comment (s-mx) [reply library/blob.c]: done

File: library/blob.c (hunk 17 above)
Comment (s-mx) [reply library/blob.c]: done

**Review by antmat — DISMISSED** [review]

File: library/blob.c (hunk 16 above)
Comment (s-mx) [reply library/blob.c]: done

File: include/eblob/blob.h (hunk 15 above)
Comment (s-mx) [reply include/eblob/blob.h:894]: done

**Review by antmat — APPROVED** [review]

**Pull request merged**